src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -550,21 +558,34 @@ private static string EntryFromPath(string entry, int offset, int length)
             }
 
             if (length == 0)
-                return String.Empty;
+                return appendPathSeparator ? PathSeparatorString : String.Empty;
 
-            // create a mutable copy
-            char[] chars = entry.ToCharArray(offset, length);
+            // Create a new resulting string instance of the appropriate substring length. The string is
+            // initially populated with PathSeparator ('/'), but will be filled in (mutated) with the actual
+            // content of the substring below, before it is returned. Path.DirectorySeparatorChar and
+            // AltDirectorySeparatorChar are skipped so that PathSeparator remains in their place, in
+            // accordance with the .ZIP File Format Specification 4.4.17.1, which states all slashes should
+            // be forward slashes. The last character remains PathSeparator if appendPathSeparator is true.
+            String result = new String(PathSeparator, appendPathSeparator ? length + 1 : length);
 
-            // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
-            // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
-            // explicitly trying to standardize to '/'
-            for(int i = 0; i < chars.Length; i++)
+            unsafe
             {
-                if (chars[i] == Path.DirectorySeparatorChar || chars[i] == Path.AltDirectorySeparatorChar)
-                    chars[i] = PathSeparator;
+                fixed (char* pResult = result)
```

Comment:
+1 to Nick's comments.  I love the desire to reduce allocations and wish there was a better way to do this but this pattern makes me really nervous.  I would prefer for us to use this as an example of while we need to solve this problem in a safe manner.

If this was a real bottleneck, the maybe we could do this.  I'd rather we just explore figuring out if we could defer this normalization until we were actually writing the entry into the zip file itself.  Ideally at that point we would not need to allocate a string at all.

In general, I like the desire to write tight, low allocation code all the time to avoid a peanut butter effect across the platform just because we don't have specific metrics on a key scenario.  However, in this case, unless we knew this was really killing us from real world profiles in main line scenarios, I think the cure is worse than the disease.